mobile_packet_verifier/src/burner.rs

```diff
@@ -0,0 +1,266 @@
+            upload_bytes: ds.upload_bytes as u64,
+            download_bytes: ds.download_bytes as u64,
+            first_timestamp: ds.first_timestamp.encode_timestamp_millis(),
+            last_timestamp: ds.last_timestamp.encode_timestamp_millis(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct PayerTotals {
+    total_bytes: u64,
+    sessions: Vec<DataTransferSession>,
+}
+
+impl PayerTotals {
+    fn push_sess(&mut self, sess: DataTransferSession) {
+        self.total_bytes += sess.download_bytes as u64 + sess.upload_bytes as u64;
+        self.sessions.push(sess);
+    }
+}
+
+pub struct Burner {
+    valid_sessions: FileSinkClient,
+    provider: RpcClient,
+    program_cache: BurnProgramCache,
+    keypair: [u8; 64],
+    cluster: String,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum BurnError {
+    #[error("Solana client error: {0}")]
+    SolanaClientError(#[from] ClientError),
+    #[error("Anchor error: {0}")]
+    AnchorError(#[from] anchor_lang::error::Error),
+    #[error("sql error: {0}")]
+    SqlError(#[from] sqlx::Error),
+    #[error("Parse pubkey error: {0}")]
+    ParsePubkeyError(#[from] ParsePubkeyError),
+    #[error("file store error: {0}")]
+    FileStoreError(#[from] file_store::Error),
+}
+
+impl Burner {
+    pub async fn new(
+        settings: &Settings,
+        valid_sessions: FileSinkClient,
+        provider: RpcClient,
+        keypair: Keypair,
+    ) -> Result<Self, BurnError> {
+        Ok(Self {
+            program_cache: BurnProgramCache::new(settings, &provider).await?,
+            provider,
+            valid_sessions,
+            keypair: keypair.to_bytes(),
+            cluster: settings.cluster.clone(),
+        })
+    }
+
+    pub async fn burn(&self, pool: &Pool<Postgres>) -> Result<(), BurnError> {
+        // Fetch all of the sessions
+        let sessions: Vec<DataTransferSession> =
+            sqlx::query_as("SELECT * FROM data_transfer_sessions")
+                .fetch_all(pool)
+                .await?;
+
+        // Fetch all of the sessions and group by the payer
+        let mut payer_totals = HashMap::<PublicKeyBinary, PayerTotals>::new();
+        for session in sessions.into_iter() {
+            payer_totals
+                .entry(session.payer.clone())
+                .or_default()
+                .push_sess(session);
+        }
+
+        for (
+            payer,
+            PayerTotals {
+                total_bytes,
+                sessions,
+            },
+        ) in payer_totals.into_iter()
+        {
+            // Fetch the sub dao epoch info:
+            let epoch = self.provider.get_epoch_info().await?.epoch;
+            let (sub_dao_epoch_info, _) = Pubkey::find_program_address(
+                &[
+                    "sub_dao_epoch_info".as_bytes(),
+                    self.program_cache.sub_dao.as_ref(),
+                    &epoch.to_le_bytes(),
+                ],
+                &helium_sub_daos::ID,
+            );
```

Comment:
do we want to do these steps for each payer in the collection of payer totals or just once for each burn session before going into the iterator?

Reply:
I thought we would have to do it individually, but we don't really. The epoch is per day, so we'll have an issue if this straddles a day